### PR TITLE
fix(aggregation): resolve grouping keys read outside the aggregation

### LINKED ISSUE #2286: `CASE` expression result changes when diagnostic projection columns are added
### Configuration

Docker Image: `falkordb/falkordb:latest`

Redis Version: `8.6.3`

FalkorDB Graph Module Version: `4.20.1`

Operating System: WSL2 - Ubuntu 24.04.1 LTS

Installation Method: Docker

---

### Steps to reproduce

1. Start FalkorDB via Docker

```text
docker run --rm --name test -p 6379:6379 falkordb/falkordb:latest
```

2. Connect with `redis-cli`

```text
docker exec -it test redis-cli
```

3. Run the following query

```text
GRAPH.QUERY test "CREATE (a:N {id:1,v:0})-[:R]->(b:N {id:2}), (b)-[:R]->(a) WITH count(*) AS ignored MATCH p=(n:N)-[:R*2..2]->(n) SET n.v = 1 WITH count(p) AS paths MATCH (n:N {v:1}) RETURN CASE WHEN paths > 0 AND count(n) > 0 THEN 'PASS' ELSE 'FAIL' END AS value"
```

4. Delete the graph and rerun the same query while also projecting the intermediate values

```text
GRAPH.DELETE test
```

```text
GRAPH.QUERY test "CREATE (a:N {id:1,v:0})-[:R]->(b:N {id:2}), (b)-[:R]->(a) WITH count(*) AS ignored MATCH p=(n:N)-[:R*2..2]->(n) SET n.v = 1 WITH count(p) AS paths MATCH (n:N {v:1}) RETURN paths, count(n) AS updated, paths > 0 AS pathPositive, count(n) > 0 AS nodePositive, paths > 0 AND count(n) > 0 AS both, CASE WHEN paths > 0 AND count(n) > 0 THEN 'PASS' ELSE 'FAIL' END AS value"
```

### Expected behaviour

Adding diagnostic projection columns must not change the value of the existing `CASE` expression. Both queries should return `PASS` because two bounded loop paths are matched and two updated nodes are found.

### Actual behaviour

The first query returns:

```text
value
FAIL
```

The second query, which differs only by projecting the intermediate expressions, returns:

```text
paths          2
updated        2
pathPositive   true
nodePositive   true
both           true
value          PASS
```

The first form returned `FAIL` in two clean-container attempts. The diagnostic form confirms that both operands are true and that projection changes the result of the same `CASE` expression.

### PR DESCRIPTION
**Pull request by olifarhaan**
## TLDR

A projection item that aggregates was evaluated against the group's key row. That row carries each grouping key only under the name the clause projects it as, so a key written in any other form was simply absent and resolved to Null, which then collapsed whatever expression contained it

The same data answered two different ways depending on nothing but the projection list:

```text
RETURN paths, paths > 0 AND count(n) > 0    ->  5, true
RETURN paths > 0 AND count(n) > 0           ->  null
```

`RETURN n.a, n.a + count(*)` did not even get that far. The projected key had taken over the slot the bare variable would have used, so it failed with `Type mismatch: expected Map, Node, Edge, ... but was Integer`

This rewrites each such sub-expression to read the key by its projected name, which the key row does carry, and rejects the reads that name a key the clause never projects. openCypher [CIP2021-07-07](https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2021-07-07-Grouping-keys-and-aggregation-expressions.adoc) requires exactly that: every sub-expression outside the aggregation must be a constant, a parameter, or a grouping key of the same clause. Neo4j reports the rest as "implicit grouping expressions". FalkorDB let them evaluate to Null, which is how the wrong answers got out

Recognized grouping keys are variables and property lookups, matching the CIP. Comprehensions, quantifiers and `reduce` are descended into carrying the variables they bind, so a key they capture resolves while their own variables are left alone. Pattern comprehensions are the one exception and are copied verbatim: the planner lifts them into their own sub-plan, which runs against the row feeding the aggregation rather than the group's key row, so the projected alias does not exist there

## User Flow

Any `WITH` or `RETURN` that aggregates while also reading a variable or property outside the aggregation. The reported shape was a `CASE`, but `CASE` is incidental, `RETURN paths + count(n)` is enough

## Relevant issues

Closes #2286

The issue attributes this to `CASE` and to "diagnostic projection columns". Neither is the mechanism, and no graph data is needed to see it

## Pre-Submission checklist

- [x] I have added meaningful tests
- [x] My PR passes all CI/CD checks (e.g., lint, format, unit tests)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem

## Screenshots / Proof of Fix

Live server, release build, module loaded into a real `redis-server`

```bash
redis-server --port 6399 --loadmodule "$PWD/target/release/libfalkordb.dylib" &
redis-cli -p 6399 GRAPH.QUERY g "CREATE (:P {a:10, b:20}), (:P {a:10, b:30})"

for q in "MATCH (n:P) RETURN n.a, n.a + count(*) AS r" \
         "UNWIND [1,2] AS n WITH 5 AS paths, n RETURN paths, paths > 0 AND count(n) > 0 AS both" \
         "MATCH (n:P) RETURN n.a + count(*) AS r" \
         "MATCH (n:P) RETURN n.a, n.b + count(*) AS r" \
         "UNWIND [1,2] AS n WITH 5 AS paths, n RETURN paths > 0 AND count(n) > 0 AS both" \
         "MATCH (n:P) WITH 0 AS i, n RETURN collect(n.a)[i] AS r"; do
  echo "\$ $q"; redis-cli -p 6399 GRAPH.QUERY g "$q" | head -4; echo
done
```

Before, on this branch's merge base:

```text
$ MATCH (n:P) RETURN n.a, n.a + count(*) AS r
Type mismatch: expected Map, Node, Edge, Datetime, Date, Time, Duration, Null, or Point but was Integer

$ UNWIND [1,2] AS n WITH 5 AS paths, n RETURN paths, paths > 0 AND count(n) > 0 AS both
paths / both -> 5 / true

$ MATCH (n:P) RETURN n.a + count(*) AS r
r -> (nil)

$ MATCH (n:P) RETURN n.a, n.b + count(*) AS r
Type mismatch: expected Map, Node, Edge, Datetime, Date, Time, Duration, Null, or Point but was Integer

$ UNWIND [1,2] AS n WITH 5 AS paths, n RETURN paths > 0 AND count(n) > 0 AS both
both -> (nil)

$ MATCH (n:P) WITH 0 AS i, n RETURN collect(n.a)[i] AS r
r -> (nil)
```

After:

```text
$ MATCH (n:P) RETURN n.a, n.a + count(*) AS r
n.a
r
10
12

$ UNWIND [1,2] AS n WITH 5 AS paths, n RETURN paths, paths > 0 AND count(n) > 0 AS both
paths
both
5
true

$ MATCH (n:P) RETURN n.a + count(*) AS r
'n.a' is read outside an aggregation function but is not a grouping key of this clause. Project 'n.a' alongside the aggregation, or compute it in a preceding WITH clause

$ MATCH (n:P) RETURN n.a, n.b + count(*) AS r
'n.b' is read outside an aggregation function but is not a grouping key of this clause. Project 'n.b' alongside the aggregation, or compute it in a preceding WITH clause

$ UNWIND [1,2] AS n WITH 5 AS paths, n RETURN paths > 0 AND count(n) > 0 AS both
'paths' is read outside an aggregation function but is not a grouping key of this clause. Project 'paths' alongside the aggregation, or compute it in a preceding WITH clause

$ MATCH (n:P) WITH 0 AS i, n RETURN collect(n.a)[i] AS r
'i' is read outside an aggregation function but is not a grouping key of this clause. Project 'i' alongside the aggregation, or compute it in a preceding WITH clause
```

The last one is Neo4j issue [#13200](https://github.com/neo4j/neo4j/issues/13200) verbatim, `with 0 as b return collect(1)[b]`, which Neo4j rejects for the same reason

Grouping keys captured by a construct that binds its own variables, which the second commit added:

```text
$ MATCH (n:P) RETURN n.a AS a, count(*) + reduce(s = 0, x IN [1] | s + n.a) AS r
a / r -> 10 / 12                    # was the same Type mismatch

$ MATCH (n:P) RETURN n.a AS a, count(*) + size([x IN [1] WHERE x < n.a | x]) AS r
a / r -> 10 / 3                     # was the same Type mismatch

$ MATCH (n:P) RETURN count(*) + reduce(s = 0, x IN [1,2] | s + x) AS r
r -> 5                              # s and x are local, untouched

$ MATCH (n:P) RETURN n.a AS a, count(*) + reduce(s = 0, x IN [1] | s + n.b) AS r
'n.b' is read outside an aggregation function but is not a grouping key of this clause. ...
```

And the issue's own query, which used to disagree with itself:

```text
$ ... WITH count(p) AS paths MATCH (n:N {v:1}) RETURN paths, count(n) AS updated, CASE ... END AS v
paths / updated / v -> 2 / 2 / PASS

$ ... WITH count(p) AS paths MATCH (n:N {v:1}) RETURN CASE ... END AS v
'paths' is read outside an aggregation function but is not a grouping key of this clause. ...
```

## Type

🐛 Bug Fix

## Changes

`graph/src/planner/binder.rs`, a set of free helpers plus one call in `bind_projection`

`tests/flow/test_aggregation.py`, three regression tests and a shared `assert_rejected` helper. That helper raises rather than calling `self.env.assertTrue(False)`, which does not reliably fail a test: an earlier draft of `test13` reported PASS on the merge base while its assertion was visibly firing

## QA runbook

All three new tests fail on the merge base and pass on this branch, with the ten pre-existing tests in the file unaffected either way

```text
# merge base
Total Tests Run: 13, Total Tests Failed: 3, Total Tests Passed: 10
	test_aggregation:testAggregations.test11_GroupingKeyReadOutsideAggregation
	test_aggregation:testAggregations.test12_ImplicitGroupingKeyIsRejected
	test_aggregation:testAggregations.test13_ProjectingAKeyDoesNotChangeOtherColumns

# this branch
Total Tests Run: 13, Total Tests Failed: 0, Total Tests Passed: 13
```

TCK is the load-bearing check for the accept side, since it pins the shape that must keep working:

```text
RELEASE=1 TCK_DONE=tck_done.txt pytest tests/tck/test_tck.py
56 features passed, 537 scenarios passed, 1 passed
```

An earlier revision of this change recognized only bare variables as grouping keys. TCK `Return6` scenario `[6] Handle aggregates inside non-aggregate expressions` caught it, which is what drove the property-lookup support. A second revision copied comprehensions, quantifiers and `reduce` verbatim, which left a captured key unresolved; that came out of review and is the second commit

Full flow suite: `Total Tests Run: 1411, Total Tests Failed: 23, Total Tests Passed: 1388`, identical before and after. `test_comprehension_functions` passes 25/25. All 23 fail inside redis-py 8.1.0 with `Redis.__init__() got an unexpected keyword argument 'himport_registry'`, which is a client-side version mismatch in my local venv and cannot be reached from the module

`tests/test_e2e.py`, `tests/test_functions.py` and `tests/test_overflow_crashes.py`: 142 passed. `cargo test -p graph`: 116 passed. `cargo fmt --all --check` clean, `CXX=clang++ cargo clippy --all-targets` exits 0 leaving only the pre-existing `too many arguments` warning in `cond_var_len_traverse.rs`

One deliberate narrowing worth flagging in review. The CIP also recognizes static map access as a grouping key. Dot access is covered, `WITH {k: 7} AS m, x RETURN m.k, m.k + count(x)` returns `7, 9`, but the bracket form `m['k']` is a different node and is rejected rather than grouped. It returned Null before, so nothing that worked stops working, and adding it later is a change to one helper

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved aggregation handling when grouping keys are used in projected expressions.
  * Preserved aggregate results and projected columns correctly during ordering and grouping.
  * Added clear errors when queries reference grouping keys that were not projected.
  * Preserved correct behavior for nested expressions and local scopes.
* **Tests**
  * Added coverage for valid grouping-key reads, rejected implicit grouping keys, and preservation of projected-column results.

### CODE REVIEW

File: graph/src/planner/binder.rs [hunk 1]
```diff
@@ -2523,6 +2525,111 @@ fn rewrite_compiled_regex(
...
+fn replace_grouping_keys(
+    node: &DynNode<ExprIR<Variable>>,
+    keys: &HashMap<GroupingKey, Variable>,
+) -> Result<DynTree<ExprIR<Variable>>, String> {
+    match node.data() {
+        ExprIR::FuncInvocation(func) if func.is_aggregate() => return Ok(node.clone_as_tree()),
+        ExprIR::ListComprehension(_)
+        | ExprIR::Quantifier { .. }
+        | ExprIR::Reduce(_)
+        | ExprIR::PatternComprehension(_) => return Ok(node.clone_as_tree()),
```
[the remark this answers is not in the document]
Comment (olifarhaan) [reply graph/src/planner/binder.rs]: Good catch, fixed in 36b2e04. Those three are evaluated inline, so a captured key now resolves. Pattern comprehensions stay verbatim since the planner evaluates them elsewhere.
